pennylane/_qubit_device.py

```diff
@@ -280,7 +280,14 @@ def generate_samples(self):
              array[complex]: array of samples in the shape ``(dev.shots, dev.num_wires)``
         """
         number_of_states = 2 ** self.num_wires
-        rotated_prob = self.probability()
+
+        # Check if device has `_analytic_probability` implemented,
+        # otherwise call `probability`
+        try:
+            rotated_prob = self._analytic_probability()
+        except AttributeError:
+            rotated_prob = self.probability()
```

Comment:
Nice!

Reply:
Really nice find!

tl;dr: we might have to enforce that statevector simulators _do_ have an `_analytic_probability` defined if they also have an `analytic` attribute. My suggestion would be to throw an error if this is not the case. HW and hw simulators will have to anyways redefine this method.

_More details_
Analytic probabilities will be needed to compute non-analytic probabilities. Although this might be coded up in a single `probability` method in general, maybe it would be best if we distinguished the two approaches from the start for plugins and required that they were separated logically.

A way to distinguish state vector simulators from HW or hw simulators is checking if the `analytic` attribute is defined.

A cleaner solution in the long rung would be creating separate classes for different types of devices (something that came up previously as well).

Reply:
If we do decide to do something like this, then we'll have to update the `QubitDevice` docs accordingly :) (Even now, it might be worth doing so at least as a note, as this is a shared part for each device that subclasses `QubitDevice`)

Reply:
I think that @josh146 mentioned that i might be quite a bit more work if making the `_analytic_probability` (and the potential `_non_analytic_probability` as you mentioned below) a requirement, especially with regards to the plugins. But maybe this is ok if only putting the requirement on simulators?

I'm guessing it's a better fit for a new PR (after some discussions, or even following an ADR).

Reply:
Having said that, this requirement would rather draw the attention of plugin writers to the fact, that this is how it works. Since this subtlety was missed in `default.qubit`, I think it would be nice if we somehow highlighted this case. Having a note in the docs could also work (if we'd rather not have a clear separation for convenience).